### ISSUE TITLE
set_index_hierarchy performance for larger datasets

### ISSUE THREAD

**Issue by naveendaftari** [InvestmentSystems/static-frame#286]
I am comparing setting a hierarchical index vs the pandas multi index, using the CRSP monthly data. I'd like to see how long it takes to set a multi-level index, by data_date and permno, in static frame and pandas. Below is some code to generate a frame/dataframe with hierarchical/multi index, as well as a timer function to time the step of setting the index. I include two versions of setting the index for static. One where I pass reorder_for_hierarchy=True to set_index_hierarchy and another version where I sort by data_date, then set the index. The timing in this setup is starkly different from pandas:

pandas ~ 1 second
static (reorder_for_hierarch=True) ~ 80 seconds
static (order dates before) ~ 50 seconds

```python
import pandas as pd
import typing as tp
import time
from papyri.api.crsp_price import CRSPPriceMonthlyDerived
import static_frame as sf

def mytimeit(func):
    def timed_method(*args, **kwargs):
        nar = time.time()
        result = func(*args, **kwargs)
        nar = round( (time.time() - nar), 2)

        if 'method' in kwargs:
            mname = kwargs['method'].__qualname__
        else: 
            mname = func.__qualname__ 

        print(f'{mname} took {nar} seconds to execute')
        return result
    
    return timed_method

@mytimeit
def set_pandas_multi_idx(un_indexed_data : pd.DataFrame, idx : tp.List[str]) -> pd.DataFrame:
    return un_indexed_data.set_index(idx)

@mytimeit
def set_static_multi_idx(un_indexed_data : sf.Frame, idx : tp.List[str]) -> sf.Frame:
    return un_indexed_data.set_index_hierarchy(idx, reorder_for_hierarchy=True, drop=True)

@mytimeit
def set_static_multi_idx_order_before(un_indexed_data : sf.Frame, idx : tp.List[str]) -> sf.Frame:
    sorted_dates = un_indexed_data.sort_values(key='data_date')
    return sorted_dates.set_index_hierarchy(idx, reorder_for_hierarchy=False, drop=True)

# get monthly pricing data from SDK and create pandas version
reader = CRSPPriceMonthlyDerived.from_latest()
data_sf = reader.read_columns(*reader.get_fields())
data_pd = data_sf.to_pandas()

# time setting of indices
index_names = ['data_date', 'permno']
pd_idx = set_pandas_multi_idx(un_indexed_data=data_pd, idx=index_names) # takes 1 second
sf_idx = set_static_multi_idx(un_indexed_data=data_sf, idx=index_names) # takes ~80 seconds
sf_idx_order_before = set_static_multi_idx_order_before(un_indexed_data=data_sf, idx=index_names) # takes ~50 seconds

**Comment by flexatone** [comment]
Possibly handled with #437

**Comment by chaburkland** [comment]
@naveendaftari This performance issue will be partially addressed by #437.

Here are the performance benchmarks I observe with your example:

```python
frame = data_sf
df = data_pd
kwargs = dict(drop=True, index_constructors=[sf.IndexDate, sf.Index])
```
### Pandas
```python
%timeit df.set_index(index_names)
455 ms ± 18.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
### New Implementation
```python
%timeit frame.set_index_hierarchy(index_names, reorder_for_hierarchy=True, **kwargs)
47.7 s ± 654 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit frame.set_index_hierarchy(index_names, reorder_for_hierarchy=False, **kwargs)
10.3 s ± 644 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
### Old Implementation
```python
%time frame.set_index_hierarchy(index_names, reorder_for_hierarchy=True, , **kwargs)
CPU times: user 1min 34s, sys: 174 ms, total: 1min 35s
Wall time: 1min 35s
```

As you can see, pandas is still much faster, but the runtime when not requiring a sort is down to ~10 seconds, and the sorted behavior is now ~50 seconds.

Additionally, the proposed new approach in #437 would make it so IndexHierarchy objects no longer fail for unsorted columns.

**Comment by naveendaftari** [comment]
Thank you for following up. This was a cool comparison to see again.
